infrastructures/infrastructure-openstack/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/OpenstackInfrastructure.java

```diff
@@ -325,4 +357,241 @@ protected void unregisterNodeAndRemoveInstanceIfNeeded(final String instanceTag,
+
+        if (!dynamicPolicyParameters.containsKey(TOTAL_NUMBER_OF_NODES_KEY)) {
+            logger.info("The dynamic policy parameters must include the total number of nodes");
+            return false;
+        } else if (!dynamicPolicyParameters.containsKey(MAX_NODES_KEY)) {
+            logger.info("The dynamic policy parameters must include the maximal number of nodes");
+            return false;
+        } else if (!dynamicPolicyParameters.containsKey(INIT_DELAY_KEY)) {
+            logger.info("The dynamic policy parameters must include the initialization delay of nodes");
+            return false;
+        } else if (!dynamicPolicyParameters.containsKey(REFRESH_TIME_KEY)) {
+            logger.info("The dynamic policy parameters must include the refresh time of the policy");
+            return false;
+        }
+
+        existingNodes = getNumberOfAcquiredNodesWithLock();
+
+        totalNodes = (Integer) dynamicPolicyParameters.get(TOTAL_NUMBER_OF_NODES_KEY);
+
+        maxNodes = (Integer) dynamicPolicyParameters.get(MAX_NODES_KEY);
+
+        initDelay = (Long) dynamicPolicyParameters.get(INIT_DELAY_KEY);
+
+        refreshTime = (Long) dynamicPolicyParameters.get(REFRESH_TIME_KEY);
+
+        if (existingNodes == totalNodes) {
+            logger.info("The number of existing nodes (" + existingNodes +
+                        ") is equal to total number of nodes required by the dynamic policy (" + totalNodes +
+                        "), so node acquisition is skipped!");
+            return false;
+        }
+
+        if (existingNodes + numberOfNodes != totalNodes) {
+            logger.warn("Dynamic policy (of Openstack infrastructure) is misconfigured. The total number of nodes (" +
+                        totalNodes + ") does not correspond to the sum of existing nodes (" + existingNodes +
+                        ") and the new nodes to add (" + numberOfNodes + ").");
+            return false;
+        }
+
+        // check that the maximal number of nodes is not reached
+        if (maxNodes - (existingNodes + numberOfNodes) < 0) {
+
+            logger.warn("The sum of existing nodes (" + existingNodes + ") and the required nodes (" + numberOfNodes +
+                        ") is greater than the maximal number of nodes (" + maxNodes +
+                        ") allowed by the dynamic policy, so deploying only " + (maxNodes - existingNodes) + " nodes.");
+
+            nbOfNodes = maxNodes - existingNodes;
+        }
+
+        return true;
+    }
+
+    private boolean checkOpenstackInstancesToDeploy() {
+
+        int existingInstances = getNumberOfInstancesWithLock();
+        logger.info("number of existing instances: " + existingInstances);
+
+        int preliminarynNumberOfInstances = (int) Math.ceil((double) nbOfNodes / (double) numberOfNodesPerInstance);
+        instancesToDeploy = Math.min(preliminarynNumberOfInstances, numberOfInstances);
```

Comment:
Why do we take Math.min ? i mean we have `preliminarynNumberOfInstances` and `numberOfInstances`, and somehow we understand how manu instances to deploy? Again, i am not saying it is wrong. But for me it is difficult to understand if it is right or not

Reply:
Evaluating `nbOfNodes / numberOfNodesPerInstance` and `nbOfNodes % numberOfNodesPerInstance`  is a way to determine `instanceToDeploy` while remaining in the integer kingdom.

Reply:
- preliminarynNumberOfInstances is determined based on the number of nodes required by the dynamic policy.
- numberOfInstances is the number of instances allowed by the infra. 
We consider the minimal number of instances the offers the number of nodes (required by the dynamic policy), and does not exceed the number of instances allowed by the infra.